appserver/web/war-util/src/test/java/org/glassfish/web/loader/WebappClassLoaderTest.java

```diff
@@ -0,0 +1,183 @@
+ * file and include the License file at glassfish/legal/LICENSE.txt.
+ *
+ * GPL Classpath Exception:
+ * The Payara Foundation designates this particular file as subject to the "Classpath"
+ * exception as provided by the Payara Foundation in the GPL Version 2 section of the License
+ * file that accompanied this code.
+ *
+ * Modifications:
+ * If applicable, add the following below the License Header, with the fields
+ * enclosed by brackets [] replaced by your own identifying information:
+ * "Portions Copyright [year] [name of copyright owner]"
+ *
+ * Contributor(s):
+ * If you wish your version of this file to be governed by only the CDDL or
+ * only the GPL Version 2, indicate your decision by adding "[Contributor]
+ * elects to include this software in this distribution under the [CDDL or GPL
+ * Version 2] license."  If you don't indicate a single choice of license, a
+ * recipient has the option to distribute your version of this file under
+ * either the CDDL, the GPL Version 2 or to extend the choice of license to
+ * its licensees as provided above.  However, if you add GPL Version 2 code
+ * and therefore, elected the GPL Version 2 license, then the option applies
+ * only if the new code is made subject to such option by the copyright
+ * holder.
+ */
+package org.glassfish.web.loader;
+
+import static org.junit.Assert.assertTrue;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+
+import org.apache.naming.resources.FileDirContext;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class WebappClassLoaderTest {
+
+    private static final int EXECUTION_COUNT = 100;
+
+    private CountDownLatch latch;
+
+    private ExecutorService executor;
+    private File junitJarFile;
+
+    @Before
+    public void setup() throws URISyntaxException {
+        // Run 3 methods at the same time, and make the pool large enough to increase
+        // the chance of a race condition
+        executor = Executors.newFixedThreadPool(60);
+
+        // Require a minimum number of executions before completing
+        latch = new CountDownLatch(EXECUTION_COUNT);
+
+        // Fetch any JAR to use for classloading
+        junitJarFile = new File(Test.class.getProtectionDomain().getCodeSource().getLocation().toURI());
+    }
+
+    @After
+    public void shutdown() throws InterruptedException {
+        if (executor != null) {
+            executor.shutdownNow();
+            assertTrue("Executor could not shutdown. This could mean there is a deadlock.",
```

Comment:
executor.awaitTermination termination always return true, also in the case the tasks isn't finished. In that case, the executor send an interrupted signal and InterruptedException occurs (verified with try catch in lookup method).